server/src/main/java/org/opensearch/common/StreamProvider.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common;
+
+import java.io.IOException;
+
+/**
+ * StreamProvider is used to supply streams to vendor plugins using <code>StreamProvider#provideStream</code>
+ */
+public class StreamProvider {
+
+    private final TransferPartStreamSupplier streamSupplier;
+    private final long partSize;
+    private final long lastPartSize;
+    private final int numOfParts;
+
+    /**
+     * Construct a new StreamProvider object
+     *
+     * @param streamSupplier An implementation of TransferPartStreamSupplier which will be called when provideStreams is called
+     * @param partSize Size of all parts apart from the last one
+     * @param lastPartSize Size of the last part
+     * @param numOfParts Total number of parts
+     */
+    public StreamProvider(TransferPartStreamSupplier streamSupplier, long partSize, long lastPartSize, int numOfParts) {
```

Comment:
This interface and `TransferPartStreamSupplier` seem to serving the same purpose of providing the stream when asked with part number. And this class contains logic of deriving the location and size basis where the requested part is the last or not. Is it possible to condense these 2 interfaces to 1?

Reply:
Added the reason of why we need both in comment above https://github.com/opensearch-project/OpenSearch/pull/7000#discussion_r1182293013.